chore(docs/halyard/service-settings): Added extensive documentation r… (#1447)
spinnaker/spinnaker.github.io: reference/halyard/custom.md

### Kubernetes

> Inside the `kubernetes` item at the top level you can add a number of custom service settings that will impact the way that the Spinnaker services are deployed onto a Kubernetes cluster. This provides you the ability to configure parts of how the Kubernetes `Deployment`, `Service`, `Pods` and other Kubernetes resources are deployed by Halyard. This allows you to do things like add a Kubernetes `secret` or `emptydir` to be mounted as a `Volume` and `VolumeMount`, change the default Kubernetes deployment strategy and more.

Reviewer: ```suggestion
Inside the `kubernetes` item at the top level you can add a number of custom service settings that  impact how the Spinnaker services are deployed onto a Kubernetes cluster. This lets you partially configure how the Kubernetes `Deployment`, `Service`, `Pods` and other Kubernetes resources are deployed by Halyard. This allows you to do things like add a Kubernetes `secret` or `emptydir` to be mounted as a `Volume` and `VolumeMount`, change the default Kubernetes deployment strategy and more.
```

> Inside the `kubernetes` item at the top level you can add a number of custom service settings that will impact the way that the Spinnaker services are deployed onto a Kubernetes cluster. This provides you the ability to configure parts of how the Kubernetes `Deployment`, `Service`, `Pods` and other Kubernetes resources are deployed by Halyard. This allows you to do things like adding a Kubernetes `secret` or `emptydir` to be mounted as a `Volume` and `VolumeMount`, change the default Kubernetes deployment strategy and more.

Reviewer 2: ```suggestion
Inside the `kubernetes` item at the top level you can add a number of custom service settings that will impact the way that the Spinnaker services are deployed onto a Kubernetes cluster. This provides you the ability to configure parts of how the Kubernetes `Deployment`, `Service`, `Pods` and other Kubernetes resources are deployed by Halyard. This allows you to do things like add a Kubernetes `secret` or `emptydir` to be mounted as a `Volume` and `VolumeMount`, change the default Kubernetes deployment strategy and more.
```

At a high level, the configurable items inside the Kubernetes service settings are:

| Item | Details |
| ----- | ----------- |

> | `deploymentStrategy` | Used to define either `RollingUpdate` or `Recreate` deployment strategy, including setting `MaxSurge` and `MaxUnavailable` See [Deployment Strategy](#deployment-strategy) below. |

Reviewer: ```suggestion
| `deploymentStrategy` | Defines either `RollingUpdate` or `Recreate` deployment strategy, including setting `MaxSurge` and `MaxUnavailable` See [Deployment Strategy](#deployment-strategy) below. |
```

Reviewer: s/Used to define/Defines/
...in all cases, because "used to" also = "fomerly"

```
kubernetes:
  podAnnotations:
    example/annotation: spinnaker.io
    example/annotation-2: halyard
  podLabels:
    examplePodLabel: examplePodLabelValue
  serviceLabels:
    exampleServiceLabel: exampleServiceLabelValue
```

These are all optional ways to control what metadata elements exist on the resources deployed by Halyard. For additional information about how these can be used, you can look at the [Annotations](https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/) or [Labels and Selectors](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/) pages from Kubernetes.

### nodeSelector

Node selector annotations can be used to put out `nodeSelector` values in the Pod specification. They generally follow a similar syntax to the annotations and labels. For example:

```
kubernetes:
  nodeSelector:
     exampleNodeKey: exampleNodeValue
```

Additional information regarding Node Selectors can be found in [the Kubernetes NodeSelector documentation](https://kubernetes.io/docs/concepts/configuration/assign-pod-node/#nodeselector).

### useExecHealthCheck

> By default halyard deploys services with an `exec` based health check in order to improve compatibility with istio. This however can break functionality for implementations of Load Balancer service types and Ingress Controllers that rely on having a http health check to validate. Setting `kubernetes.useExecHealthCheck: false` will switch the check method to be http based for such use cases.

Reviewer: ```suggestion
By default halyard deploys services with an `exec`-based health check in order to improve compatibility with Istio. This however can break functionality for implementations of Load Balancer service types and Ingress Controllers that rely on having an http health check to validate. Setting `kubernetes.useExecHealthCheck: false` will switch the check method to be http based for such use cases.
```

### Using Custom Volumes

You can define a set of custom volumes for a particular service that will become `Volume` and `VolumeMount` items in the generated pod specifications when Halyard deploys Spinnaker on Kubernetes. This can be useful if you need to add custom CA Certificates or other files to Spinnaker service consistently.

The Kubernetes service settings allow you to mount volumes of type `configMap`, `secret` and `emptyDir` with the below syntax.

```
kubernetes:
  volumes:
  - id: secret-example-vol
    type: secret
    mountPath: /path/to/location/of/secret
  - id: configmap-example-vol
    type: configMap
    mountPath: /path/to/location/of/configMap
  - id: emptydir-example-vol
    type: emptyDir
    mountPath: /path/to/location/of/emptyDir
```
Make sure that the `id` field matches the `name` of the existing `secret` or `configmap` resource in Kubernetes that you need mounted.

### Deployment Strategy

> You can change the Deployment Strategy that is used when deploying Spinnaker services.

Reviewer: ```suggestion
You can change the deployment strategy that is used when deploying Spinnaker services.
```

For example, if you wanted to adjust the `maxSurge` and `maxUnavailable` percentages for a rolling update of Clouddriver, you would define a `clouddriver.yml` file in your Halyard `service-settings` like the one below.

```
kubernetes:
  deploymentStrategy:
    type: rollingUpdate
    strategySettings:
      maxSurge: 100%
      maxUnavailable: 50%
```

If you wanted to use a `Recreate` strategy instead, you would use:

```
kubernetes:
  deploymentStrategy:
    type: recreate
```

> For additional information regarding the implication of each of these Deployment Strategies, consult the [Updating a Deployment](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/) documentation.

Reviewer: ```suggestion
For additional information regarding the implication of each of these deployment strategies, consult the [Updating a Deployment](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/) documentation.
```
